ocp_resources/datavolume.py

```diff
@@ -269,29 +274,73 @@ def _check_none_pending_status(self, failure_timeout=120):
-            self.logger.error(f"{self.name} status is {sample}")
+            self.logger.error(f"{self.name} status is {sample.status}")
             raise
 
-    def wait_for_dv_success(self, timeout=600, failure_timeout=120):
+    def _check_dv_progress(self):
+        dv_status = self.instance.status
+        dv_status_conditions = dv_status.get("conditions")
+        if (
+            dv_status.restartCount
+            and dv_status.restartCount > 3
+            and dv_status_conditions
+            and dv_status_conditions[-1].type == "Running"
+            and dv_status_conditions[-1].status == "False"
+        ):
+            raise TimeoutExpiredError(
+                value=(
+                    f"Status of {self.kind} '{self.name}' in '{self.namespace}' namespace is \n"
+                    f"{dv_status_conditions}"
+                )
+            )
+
+    def wait_for_dv_success(
+        self,
+        timeout=TIMEOUT_10MINUTES,
+        failure_timeout=TIMEOUT_2MINUTES,
+        check_progress=False,
+    ):
+        """
+        Wait until DataVolume succeeded with or without DV Garbage Collection enabled
+
+        Args:
+            timeout (int):  Time to wait for the DataVolume to succeed.
+            failure_timeout (int): Time to wait for the DataVolume to have not Pending/None status
+            check_progress (bool, default: False): True to raise TimeoutExpiredError if DataVolume restartCount is >3
+                and one of the DV conditions is "Running": "False"
+
+        Returns:
+            bool: True if DataVolume succeeded.
+        """
         self.logger.info(f"Wait DV success for {timeout} seconds")
         self._check_none_pending_status(failure_timeout=failure_timeout)
-        self.pvc.wait_for_status(
-            status=PersistentVolumeClaim.Status.BOUND, timeout=timeout
-        )
+
         # if garbage collector enabled, the DV deleted once succeeded
+        sample = None
         try:
             for sample in TimeoutSampler(
                 sleep=1,
                 wait_timeout=timeout,
                 func=lambda: self.exists,
             ):
-                # DV reach to success if the status is succeeded or if the DV does not exists
+                # DV reach to success if the status is succeeded or if the DV does not exist
                 if sample is None or sample.status.phase == self.Status.SUCCEEDED:
-                    return
+                    break
+                elif check_progress:
+                    self._check_dv_progress()
+
         except TimeoutExpiredError:
-            self.logger.error(f"Status of {self.kind} {self.name} is {sample}")
+            self.logger.error(
+                f"Status of {self.kind} '{self.name}' in '{self.namespace}' namespace is \n"
+                f"{sample.status}"
+            )
             raise
 
+        # For CSI storage, PVC gets Bound after DV succeeded
+        return self.pvc.wait_for_status(
```

Comment:
this return don't belong to this function which only wait for dv (`wait_for_dv_success`)

Reply:
we can't say DV succeeded if PVC is not Bound.
And if DV was garbage collected - Bounded PVC is our only evidence.